Surface SES on globalThis

kriskowal opened the pull request:
Previously, SES exported `lockdown`. Calling `lockdown` introduces a global `harden`.

This change removes all exports and instead surfaces the same names on `globalThis`. That is, `lockdown`, `harden`, `Compartment`, and `ModuleStaticRecord` are all globals as they would be if they were introduced by the platform rather than a shim.

Fixes #301

erights commented on packages/ses/src/lockdown-shim.js:
```diff
@@ -124,16 +112,22 @@ export function lockdown(options = {}) {
   // must be the operation that modifies the intrinsics.
-  harden(intrinsics, registerOnly);
-  harden(detachedProperties, registerOnly);
+  lockdownHarden(intrinsics, registerOnly);
```
Where is this `harden` (aka `lockdownHarden`) function that takes a second parameter? I couldn't find it.

kriskowal replied:
https://github.com/Agoric/SES-shim/blob/0e83a286dc01ee16d20434548bd8bac0faab30bc/packages/ses/src/lockdown-shim.js#L42

erights replied:
Whose source is at https://github.com/Agoric/SES-shim/blob/0e83a286dc01ee16d20434548bd8bac0faab30bc/packages/make-hardener/src/main.js#L67 which takes only one parameter?

kriskowal replied:
I checked. `lockdownHarden`, which is just the local name for `harden` that is usable before `lockdown` finishes, does not accept a second argument. To me, this looks like the `registerOnly` behavior does not work and (because tests pass) is not tested. I don’t know what it is meant to do.

erights replied:
What is `registerOnly` supposed to mean or do? What is it needed for?

kriskowal replied:
There is no indication anywhere in the code or documentation of what `registerOnly` means or should do. I would be glad to erase it in this change.

erights replied:
Yes, please do.

For a small security kernel, I'd prefer to err on the side of removing things we don't know the purpose of. Sometimes, in review, I'll know and clarify.

erights commented on packages/ses/src/main.js:
```diff
@@ -12,5 +12,12 @@
+import { Compartment, ModuleStaticRecord } from './compartment-shim.js';
+
+Object.assign(globalThis, {
+  lockdown,
+  harden,
+  Compartment,
+  ModuleStaticRecord,
```
All shimmed pervasively global names must be added to the appropriate whitelists. I don't see `lockdown` and `ModuleStaticRecord` (or `StaticModuleRecord`) in `intrinsic-names.js`

I do see `ModuleStaticRecord` in `intrinsic-globals.js`, but I don't see `lockdown` there.

`lockdown` missing from `whitelist.js` as well.

We should have a test that fails on such omission, because it is an easy mistake to keep making. 

TODO(markm): the whole intrinsic management system now consists of nine separate files with too much redundancy, e.g., between `intrinsic-names.js`, `intrinsic-globals.js`, and `whitelist.js`. We should simplify, make it easier to maintain, and self-test consistency among what redundancy remains. Also, `check-intrinsics.js` seems vacuous. Bad edit from a previously meaningful state?

erights replied:
Filed TODO as https://github.com/Agoric/SES-shim/issues/317

kriskowal replied:
This is the kind of thing that I worry most about. Thanks for catching this.

kriskowal replied:
Adding `lockdown` to the whitelist caused tests to fail, so I’ll come back to it when we have a firm resolution as to whether it should preserve itself after `lockdown`.

erights replied:
ok. But make sure the state as of merging this PR is consistent with itself.

erights commented on packages/ses/src/lockdown-shim.js:
```diff
@@ -35,6 +33,19 @@ function assert(condition, message) {
   }
```
Why not import this from `assert.js`? It is currently identical, and it will make it easy to upgrade all occurrences to the new assert system when it is time to do that.

erights commented on packages/ses/README.md:
```diff
@@ -27,34 +27,79 @@ npm install ses
+`globalThis` and wholly independent system of modules, but otherwise shares
+the same batch of intrinsics like `Array` with the surrounding compartment.
+The concept of a compartment implies the existence of a "start compartment",
+the initial execution argument.
```
I don't think you mean "argument"

erights commented on packages/ses/README.md:
```diff
@@ -27,34 +27,79 @@ npm install ses
 
 The new compartment has a different global object than the start compartment.
 The global object is initially mutable.
 Locking down the start compartment hardened many of the intrinsics in global
```
Avoid the phrase "locking down the start compartment". This invites the confusion the rest of the paragraph corrects. `lockdown` locks down the realm, and therefore the shared intrinsics, that the start compartment shares with these other compartments.

erights commented on packages/ses/README.md:
````diff
@@ -27,34 +27,79 @@ npm install ses
+After `lockdown`, no compartment can tamper with these intrinsics.
 Many of these intrinsics are identical in the new compartment.
 
 ```js
````
Avoid terms like "the property holds" below when you're not referring to object properties but in a context where you might be.

erights commented on packages/ses/README.md:
```diff
@@ -27,34 +27,79 @@ npm install ses
-The compartment can import modules and evaluate programs.
+SES introduces the `lockdown` function.
+Alters the surrounding execution environment such that no object intrinsically
+accessible, like the array or function prototypes (`[].prototype` or `(() =>
+0).prototype`, can be replaced or subverted to attack other code in the same
```
Thread is at https://github.com/tc39/how-we-work/pull/84

warner commented on packages/ses/README.md:
```diff
@@ -27,34 +27,79 @@ npm install ses
-that that object and any object reachable from its surface.
-The compartment can import modules and evaluate programs.
+SES introduces the `lockdown` function.
+Alters the surrounding execution environment such that no object intrinsically
```
grammar/readability nits: maybe:

* Calling `lockdown` alters the surrounding..
* .. no intrinsically-accessible object, like ..
* . `lockdown` also tames other ..

warner commented on packages/ses/README.md:
````diff
@@ -27,34 +27,79 @@ npm install ses
 
 ```js
-import {lockdown} from "ses";
+import 'ses';
+
+import 'my-vetted-shim';
+
+lockdown();
````
Hm, that shows off the simple one-file use case well. I'm on the fence as to whether `my-vetted-shim` should be imported before `ses` (to visually suggest that it is unconfined), or after (to remind readers that nothing is actually confined until `lockdown()` is called).

But a more realistic usage summary would involve confined modules too. With the current feature set, that means writing an `install-ses` module (consisting entirely of `import 'ses'; lockdown();`), and doing the other imports before/after it:

```js
import 'my-vetted-shim';
import 'install-ses';
import 'confined-module';
other_confined_code();
```
Do we want to show this here/now, or will the upcoming module-import support provide a better way? I'm hoping we'll have some docs that explain how to use this in a way similar to how folks are used to writing/running programs today, rather than putting off the practicalities in favor of a future application-loader/manifest/"larger-scale" ecosystem that we're still trying to figure out.

erights replied:
`my-vetted-shim` is ***not*** confined; which is why it, along with any code run before `lockdown()`, must be vetted. IOW, all code run before `lockdown()` can do powerful dangerous things, creating unlimited damage to everything run afterward, including `lockdown()` itself.

Running before `import 'ses';` vs between it and `lockdown()` makes only one difference: whether `harden`,`ModuleStaticRecord`,`Compartment`, and `lockdown` are available as globals. 

I just looked at the shim source and confirmed that `import 'ses';` by itself causes *no* effects other than making these available as properties on the global object and thereby as global variables.

kriskowal replied:
I believe @warner is speaking to the point of an `install-ses` module that looks like:

```js
import 'ses';
lockdown();
```

Such that in his example, `confined-module` is running after lockdown.

My intuition is that this is dangerously clever since the so-called confined module still has access to power that lockdown was unable to expunge from the start compartment.

Not sure whether we should highlight that usage pattern, as I am uncertain about how dangerous it is.

warner commented on packages/ses/README.md:
```diff
@@ -74,8 +119,6 @@ c1.global === c2.global; // false
 c1.global.JSON === c2.global.JSON; // true
```
`c1.global` or `c1.globalThis`? I guess it depends upon whether #313 lands before or after this one.

kriskowal replied:
This is the PR that resolves #313, so I’ll fix this here!

kriskowal commented on packages/ses/README.md:
```diff
@@ -27,34 +27,79 @@ npm install ses
+
+SES introduces the `harden` function.
+*After* calling `lockdown`, the `harden` function ensures that whatever object
+you give it is also transitively frozen out to the execution environments
```
We have no explanation, not even inline. https://github.com/Agoric/SES-shim/issues/320

kriskowal commented:
I’ve responded to most the great feedback from @warner and @erights and request review to the two new fixup commits I’ve added to the end.

* Lots of refinement on the verbiage.
* Removed the `registerOnly` option that didn’t do anything.
* Left `lockdown` out of the whitelists, but made sure `ModuleStaticRecord` and `Compartment` got in there with their prototypes, and `harden`.

erights commented on packages/ses/README.md:
```diff
@@ -29,103 +29,121 @@ npm install ses
+realm.
+The set of accessible objects includes but is not limited to: `globalThis`,
+`[].__proto__`, `{}.__proto__`, `(() => {}).__proto__` `(async () =>
+{}).__proto__`, the properties of any accessible object, the return values of
+any accessible function, and the result of using any accessible constructor.
```
> the return values of any accessible function, and the result of using any accessible constructor.

Either I misunderstand you, or this is wrong. Calling a function or constructor can, and often does, allocate and return a fresh mutable object not locked down by `lockdown()`. I think just deleting that part of the sentence fixes the problem.
